packages/x-internals/src/fastShallowCompare/fastShallowCompare.ts

```diff
@@ -0,0 +1,19 @@
+import { fastArrayCompare } from '../fastArrayCompare';
+import { fastObjectShallowCompare } from '../fastObjectShallowCompare';
+
+/**
+ * Fast shallow compare.
+ *
+ * Builds on top of `fastArrayCompare` and `fastObjectShallowCompare` and uses them for their respective types.
+ *
+ * @returns true if `a` and `b` are equal.
+ */
+export function fastShallowCompare(a: any, b: any) {
+  if (Array.isArray(a)) {
+    return fastArrayCompare(a, b);
+  }
+  if (b instanceof Object) {
+    return fastObjectShallowCompare(a, b);
+  }
```

Comment:
This logic is a bit confusing to me, why is it checking `a` for array and `b` for object?

Also the object check should probably be `typeof x === 'object' && x !== null`. `instanceof Object` will fail for `Object.create(null)`, which is required in some cases (e.g. inserting entries by id in an object where the id is a string provided by the user - otherwise ids like `'constructor'` or `'__proto__'` could mess up things).

In general I wouldn't advise using the "fast" methods as generic equality check functions, the tradeoff for "fast" is usually correctness in edge-cases. Imo we should maybe even mark those functions as `unsafe_fastObjectShallowCompare` to clearly indicate there are invariants that must be taken care of by the caller.

Reply:
> This logic is a bit confusing to me, why is it checking `a` for array and `b` for object?

Mistake

> In general I wouldn't advise using the "fast" methods as generic equality check functions, the tradeoff for "fast" is usually correctness in edge-cases. Imo we should maybe even mark those functions as `unsafe_fastObjectShallowCompare` to clearly indicate there are invariants that must be taken care of by the caller.

Should we use a "complete compare" instead?

Reply:
Depends on the use-case. Maybe I should have written more in the jsdoc, but the fast object compare is only really meant for record objects like `{ a: 1, b: 2, ... }`, and will possibly fail with anything else (e.g. prototype different from `Object`, a `Date`, a `RegExp`, non-enumerable props, etc). If you can guarantee that it will not receive non-exotic objects, then it's safe to use, but otherwise you should indeed use a "complete compare" focused on robustness/correctness.

Reply:
Also note that when I used `fastObjectShallowCompare` for our selectors in the grid, it was to deal with mutable records that were kept in a ref - in other words, it was a workaround for a hack. Moving forward, I plan that we don't use it anymore for selectors.

Imho, selectors should receive standard arguments, not bag-of-data arguments like `{ seriesId, etc }`, then the standard args memoization logic will work. It will also improve performance to avoid those allocations.